### PR TITLE
[WIP] Entity Rework

### PR DESCRIPTION
**Pull request by LeoDog896**
Entity UUIDs are only used in specific use scenarios and not storing them is a good way to save a lot of Ram.

### CODE REVIEW

File: src/main/java/net/minestom/server/entity/Player.java [hunk 1]
```diff
@@ -195,8 +195,10 @@ public static int getPlayerSynchronizationTickDelay(int viewersCount) {
     // Tick related
     private final PlayerTickEvent playerTickEvent = new PlayerTickEvent(this);
 
+    protected UUID uuid;
+
     public Player(@NotNull UUID uuid, @NotNull String username, @NotNull PlayerConnection playerConnection) {
-        super(EntityType.PLAYER, uuid);
+        super(EntityType.PLAYER);
         this.username = username;
```
Comment (jglrxavpok) [inline src/main/java/net/minestom/server/entity/Player.java:202]: The player's UUID is not set

Comment (LeoDog896) [reply src/main/java/net/minestom/server/entity/Player.java:202]: Good point!

**Comment by TheMode** [comment]
I don't really get the advantage, did you profile the memory usage difference?
For me it does a lot more harm than good

File: src/main/java/net/minestom/server/entity/Entity.java [hunk 2]
```diff
@@ -531,7 +519,7 @@ public void tick(long time) {
                     this.velocity.setX(velocity.getX() * drag);
                     this.velocity.setZ(velocity.getZ() * drag);
 
-                    if (velocity.equals(new Vector())) {
```
Comment (TheMode) [inline src/main/java/net/minestom/server/entity/Entity.java]: This line is required to compare against Vector#getEpsilon

File: src/main/java/net/minestom/server/entity/Entity.java [hunk 3]
```diff
@@ -181,17 +156,18 @@ public static Entity getEntity(int id) {
...
-     * @param uuid the entity UUID
-     * @return the entity having the specified uuid, null if not found
+     * @param uuid the entity unique universal id
+     * @return the entity having the specified id, null if not found
      */
     @Nullable
-    public static Entity getEntity(@NotNull UUID uuid) {
-        return Entity.entityByUuid.getOrDefault(uuid, null);
+    public static Entity getEntity(UUID uuid) {
+        return Entity.entityById.getOrDefault((int) uuid.getLeastSignificantBits(), null);
```
Comment (TheMode) [inline src/main/java/net/minestom/server/entity/Entity.java:168]: I am personally still not ok with this, it does add a bit of confusion by giving Entity and Player a different meaning of UUID

**Review by TheMode — CHANGES_REQUESTED** [review]

File: src/main/java/net/minestom/server/entity/Entity.java (hunk 3 above)
Comment (LeoDog896) [reply src/main/java/net/minestom/server/entity/Entity.java:168]: That really _is_ the point -- An Entity's UUID is **not** the same as a Player's UUID.

Comment (TheMode) [reply src/main/java/net/minestom/server/entity/Entity.java:168]: What is the difference? Without it you cannot save any entity persistently, cannot change UUID after initialization, and `Entity#getEntity(UUID)` will not work for players

**Comment by LeoDog896** [comment]
Any insight into the gravity issue?

**Pull request closed without merging**